bluetoe/link_layer/include/bluetoe/peripheral_latency.hpp

```diff
@@ -0,0 +1,318 @@
+#ifndef BLUETOE_LINK_LAYER_PERIPHERAL_LATENCY_HPP
+#define BLUETOE_LINK_LAYER_PERIPHERAL_LATENCY_HPP
+
+#include <bluetoe/ll_meta_types.hpp>
+
+/**
+ * @file bluetoe/peripheral_latency.hpp
+ *
+ * This file provides several options, how Bluetoe will implement peripheral latency. Peripheral latency
+ * allows a link layer peripheral to miss some connection events (do not respond to the central) to
+ * conserve power.
+ *
+ * Each option is a compromise between latency and power consumption.
+ *
+ * @sa bluetoe::link_layer::peripheral_latency_ignored
+ * @sa ...
+ */
+namespace bluetoe {
+namespace link_layer {
+
+    namespace details {
+        struct peripheral_latency_meta_type {};
+    }
+
+    /**
+     * @brief detailed options for the peripheral latency behavior
+     *
+     * Every option defines a set of events / circumstances under which
+     * the link layer should listen to the very next connection event.
+     * (beside the minumum events, the link layer has to listen to due
+     * to the periperhal latency value of the connection).
+     *
+     * @note in combination, there are some options that make no sense.
+     *       For example, using listen_always with some of the other
+     *       listen_* options is pointless. Also combining send_always
+     *       with send_pending_data doesn not make sense.
+     */
+    enum class peripheral_latency
+    {
+        /**
+         * @brief listen at all connection events.
+         *
+         * This does not imply, that the link layer will also respond at every
+         * connection event.
+         */
+        listen_always,
+
+        /**
+         * @brief listen, if the link layer contains unacknowlaged data.
+         *
+         * If a PDU was send out, but no acknowlagement was jet received,
+         * the link layer will listen at the next connection event, if
+         * this options is set.
+         *
+         * This might be interesting, if the sending queue is very small because
+         * Bluetoe has to keep a copy of a transmitted PDU, as long as that PDU
+         * is not acknowlaged.
+         */
+        listen_if_unacknowlaged_data,
+
+        /**
+         * @brief listen to the next connection event, if the last received PDU was
```

Comment:
As discussed on chat, maybe listen for N connection events after receiving a non-empty PDU. But I'm not totally convinced about this, what about this becomes a 'runtime setting' (using listen always) and this gets controlled by the application, rather than the stack ?

Reply:
The case which we had identified for this was "keep listening at immediately following connection event if received some data (_even if there is no MD bit - could be that the central can only send one packet per connection event due to scheduling constraints_)"

I.e., in fact it seems that would make sense only for a central that did not implement the More Data, or a for a central that is configured to have just one Tx buffer.

So, I agree we can probably ditch that one, because a reasonably designed central should implement the MD bit, which this is the purpose of.

---

And yes this is also case N == 1 of the mechanism we discussed yesterday and now I tend to agree that this should probably be driven by the application. Maybe using something like the 'override' mechanism that was evocated in https://docs.google.com/spreadsheets/d/1oM0Sh0aebYOlET8mX8RIwza3NqQsD99LcxVKeeCh5E8/edit#gid=478589848&range=C46 - could be a later improvement.